src/profiling/profiler.cpp

```diff
@@ -1,442 +1,401 @@
-#include <numeric>
+#include <cstdio>
+#include <iostream>
 
-#ifdef ARB_HAVE_GPU
-    #include <cuda_profiler_api.h>
-#endif
+#include <util/span.hpp>
+#include <util/rangeutil.hpp>
 
-#include <common_types.hpp>
-#include <communication/global_policy.hpp>
-#include <profiling/profiler.hpp>
-#include <util/make_unique.hpp>
-#include <util/debug.hpp>
+#include "profiler.hpp"
 
 namespace arb {
 namespace util {
 
-// Here we provide functionality that the profiler can use to control the CUDA
-// profiler nvprof. The start_nvprof and stop_nvprof calls are provided to let
-// a program control which parts of the program are to be profiled. It is a
-// simple wrapper around the API calls with a mutex to ensure correct behaviour
-// when multiple threads attempt to start or stop the profiler.
-#ifdef ARB_HAVE_GPU
-namespace gpu {
-    bool is_running_nvprof = false;
-    std::mutex gpu_profiler_mutex;
-
-    void start_nvprof() {
-        std::lock_guard<std::mutex> guard(gpu_profiler_mutex);
-        if (!is_running_nvprof) {
-            cudaProfilerStart();
-        }
-        is_running_nvprof = true;
+#ifdef ARB_HAVE_PROFILING
+namespace {
+    // Check whether a string describes a valid profiler region name.
+    bool is_valid_region_string(const std::string& s) {
+        if (s.size()==0u || s.front()=='_' || s.back()=='_') return false;
+        return s.find("__") == s.npos;
     }
 
-    void stop_nvprof() {
-        std::lock_guard<std::mutex> guard(gpu_profiler_mutex);
-        if (is_running_nvprof) {
-            cudaProfilerStop();
+    //
+    // Return a list of the words in the string, using '_' as the delimiter
+    // string, e.g.:
+    //      "communicator"             -> {"communicator"}
+    //      "communicator_events"      -> {"communicator", "events"}
+    //      "communicator_events_sort" -> {"communicator", "events", "sort"}
+    std::vector<std::string> split(const std::string& str) {
+        std::vector<std::string> cont;
+        std::size_t first = 0;
+        std::size_t last = str.find('_');
+        while (last != std::string::npos) {
+            cont.push_back(str.substr(first, last - first));
+            first = last + 1;
+            last = str.find('_', first);
         }
-        is_running_nvprof = false;
+        cont.push_back(str.substr(first, last - first));
+        return cont;
     }
```

Comment:
An alternative would be to make a `std::istringstream` on `str` and use `std::getline()` with the delimiter — no more efficient, but it might read more simply.

Reply:
This works, and isn't too complicated!